File: src/main/java/fr/openent/pmb/services/SchoolService.java
package fr.openent.pmb.services;

import fr.wseduc.webutils.Either;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.entcore.common.user.UserInfos;

import java.util.List;

public interface SchoolService {
    /**
     * List all schools in neo4j
     * @param handler function handler returning JsonArray data
     */
    void listNeo(Handler<Either<String, JsonArray>> handler);

    /**
     * List all schools
     * @param handler function handler returning JsonArray data
     */
    void list(Handler<Either<String, JsonArray>> handler);

    /**
     * Create a school
     * @param schools JsonArray data
     * @param handler function handler returning JsonArray data
     */
    void create(JsonArray schools, Handler<Either<String, JsonArray>> handler);

    /**
     * Update the PMB connection of a specific school (host/endpoint/source_id/credentials/
     * page_size) without touching its idneo/uai/nom/principal fields.
     * @param schoolId school identifier
     * @param connection JsonObject data (pmbHost, pmbEndpoint, pmbSourceId, pmbUsername, pmbPassword, pmbPageSize)
     * @param handler function handler returning JsonObject data
     */
    void updateConnection(String schoolId, JsonObject connection, Handler<Either<String, JsonObject>> handler);

    /**
     * Delete a scpecific school
     * @param schoolId school identifier
     * @param handler function handler returning JsonObject data
     */
    void delete(String schoolId, Handler<Either<String, JsonObject>> handler);

    /**
     * Get UAIs from schools in a city
     * @param uai school's uai
     * @param handler function handler returning JsonArray data
     */
    void getCity(String uai, Handler<Either<String, JsonArray>> handler);

}

File: src/main/resources/public/ts/services/SchoolService.ts
import {idiom, ng, notify} from 'entcore';
import http, {AxiosResponse} from 'axios';
import {School} from '../models';

export interface SchoolConnection {
    pmbHost: string;
    pmbEndpoint: string;
    pmbSourceId: string;
    pmbUsername: string;
    pmbPassword: string;
    pmbPageSize: number;
}

export interface SchoolService {
    list() : Promise<AxiosResponse>;
    listNeo() : Promise<AxiosResponse>;
    create(schools: School[]) : Promise<AxiosResponse>;
    updateConnection(schoolId: number, connection: SchoolConnection) : Promise<AxiosResponse>;
    delete(schoolId: number) : Promise<AxiosResponse>;
}

export const schoolService: SchoolService = {

    async list() : Promise<AxiosResponse> {
        try {
            return http.get('/pmb/schools');
        } catch (err) {
            notify.error(idiom.translate('pmb.error.schoolService.list'));
            throw err;
        }
    },

    async listNeo() : Promise<AxiosResponse> {
        try {
            return http.get('/pmb/schools/neo');
        } catch (err) {
            notify.error(idiom.translate('pmb.error.schoolService.list'));
            throw err;
        }
    },

    async create(schools: School[]) : Promise<AxiosResponse> {
        try {
            return http.post('/pmb/schools', schools);
        } catch (err) {
            notify.error(idiom.translate('pmb.error.schoolService.create'));
            throw err;
        }
    },

    async updateConnection(schoolId: number, connection: SchoolConnection) : Promise<AxiosResponse> {
        try {
            return await http.put(`/pmb/schools/${schoolId}/connection`, connection);
        } catch (err) {
            notify.error(idiom.translate('pmb.error.schoolService.updateConnection'));
            throw err;
        }
    },

    async delete(schoolId: number) : Promise<AxiosResponse> {
        try {
            return await http.delete(`/pmb/schools/${schoolId}`);
        } catch (err) {
            notify.error(idiom.translate('pmb.error.schoolService.delete'));
            throw err;
        }
    }

};

export const SchoolService = ng.service('SchoolService', (): SchoolService => schoolService);